File: scripts/generate-icons.py
#!/usr/bin/env python3
"""Generate the Clork app icon for each theme.

Design (matches assets/images/icon.png): solid accent background, thick ink
clock (ring stroke ~16% of radius, hands at 10:08 forming a check mark,
center dot) plus three rounded planning lines on the right. The mark is
centered-left (~40% x). Ink color = theme `onAccent` (light ink on dark
accents like plum/graphite).

Outputs:
  assets/icons/icon-<themeId>.png  (1024x1024, one per theme)
  assets/images/icon.png           (regenerated default = honey)

Keep palettes in sync with src/constants/themes.ts.

Usage: python3 scripts/generate-icons.py
Requires: Pillow (python3 -m pip install --user pillow)
"""

from __future__ import annotations

import math
from pathlib import Path

from PIL import Image, ImageDraw

ROOT = Path(__file__).resolve().parent.parent
ICONS_DIR = ROOT / "assets" / "icons"
DEFAULT_ICON = ROOT / "assets" / "images" / "icon.png"

# (accent, onAccent) — sync with src/constants/themes.ts
THEMES: dict[str, tuple[str, str]] = {
    "honey": ("#FFC233", "#26210E"),
    "blossom": ("#F9A8C9", "#2E1220"),
    "plum": ("#7C5CB8", "#FFF8F2"),
    "water": ("#7EC8E3", "#0E2530"),
    "sage": ("#9CC5A1", "#14261A"),
    "graphite": ("#4A4A52", "#F4F2E8"),
}

SIZE = 1024  # final output size
SS = 4  # supersampling factor for crisp anti-aliased edges

# Geometry measured on the original honey icon (1024x1024).
CLOCK_CENTER = (415, 500)  # ~40% x, mark centered-left
CLOCK_RADIUS = 268
RING_STROKE = 46  # ~17% of radius
HAND_STROKE = 42
CENTER_DOT_RADIUS = 30
# 10:08 — minute hand up-right, hour hand up-left, together forming a check.
MINUTE_ANGLE_DEG = 48  # clockwise from 12 o'clock
MINUTE_LENGTH = 192
HOUR_ANGLE_DEG = 304
HOUR_LENGTH = 134

LINE_STROKE = 38
LINE_X_START = 740
PLANNING_LINES = (  # (y_center, x_end) — shrinking widths, right block
    (432, 1005),
    (510, 925),
    (588, 850),
)


def rounded_line(
    draw: ImageDraw.ImageDraw,
    start: tuple[float, float],
    end: tuple[float, float],
    stroke: float,
    color: str,
) -> None:
    """Line with round caps."""
    draw.line([start, end], fill=color, width=round(stroke))
    half = stroke / 2
    for x, y in (start, end):
        draw.ellipse([x - half, y - half, x + half, y + half], fill=color)


def polar_offset(
    origin: tuple[float, float], angle_deg: float, length: float
) -> tuple[float, float]:
    """Point at `length` from origin, angle measured clockwise from 12 o'clock."""
    rad = math.radians(angle_deg)
    return (origin[0] + math.sin(rad) * length, origin[1] - math.cos(rad) * length)


def draw_icon(accent: str, ink: str) -> Image.Image:
    size = SIZE * SS
    image = Image.new("RGB", (size, size), accent)
    draw = ImageDraw.Draw(image)

    cx, cy = (CLOCK_CENTER[0] * SS, CLOCK_CENTER[1] * SS)
    radius = CLOCK_RADIUS * SS

    # Clock ring.
    draw.ellipse(
        [cx - radius, cy - radius, cx + radius, cy + radius],
        outline=ink,
        width=RING_STROKE * SS,
    )

    # Hands at 10:08 (check mark).
    for angle, length in ((MINUTE_ANGLE_DEG, MINUTE_LENGTH), (HOUR_ANGLE_DEG, HOUR_LENGTH)):
        tip = polar_offset((cx, cy), angle, length * SS)
        rounded_line(draw, (cx, cy), tip, HAND_STROKE * SS, ink)

    # Center dot.
    dot = CENTER_DOT_RADIUS * SS
    draw.ellipse([cx - dot, cy - dot, cx + dot, cy + dot], fill=ink)

    # Planning lines (right block).
    for y_center, x_end in PLANNING_LINES:
        rounded_line(
            draw,
            (LINE_X_START * SS, y_center * SS),
            (x_end * SS, y_center * SS),
            LINE_STROKE * SS,
            ink,
        )

    return image.resize((SIZE, SIZE), Image.LANCZOS)


def main() -> None:
    ICONS_DIR.mkdir(parents=True, exist_ok=True)
    for theme_id, (accent, ink) in THEMES.items():
        icon = draw_icon(accent, ink)
        target = ICONS_DIR / f"icon-{theme_id}.png"
        icon.save(target, format="PNG")
        print(f"wrote {target.relative_to(ROOT)}")

    # Default app icon = honey.
    accent, ink = THEMES["honey"]
    draw_icon(accent, ink).save(DEFAULT_ICON, format="PNG")
    print(f"wrote {DEFAULT_ICON.relative_to(ROOT)} (default, honey)")


if __name__ == "__main__":
    main()

File: scripts/generate-icons.mjs
#!/usr/bin/env node
// Node fallback of scripts/generate-icons.py (same design, same outputs) for
// environments without Python/Pillow. Pure Node (zlib) — no dependency.
// Usage: node scripts/generate-icons.mjs

import { deflateSync } from "node:zlib";
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const ICONS_DIR = join(ROOT, "assets", "icons");
const DEFAULT_ICON = join(ROOT, "assets", "images", "icon.png");

// [accent, onAccent] — sync with src/constants/themes.ts
const THEMES = {
  honey: ["#FFC233", "#26210E"],
  blossom: ["#F9A8C9", "#2E1220"],
  plum: ["#7C5CB8", "#FFF8F2"],
  water: ["#7EC8E3", "#0E2530"],
  sage: ["#9CC5A1", "#14261A"],
  graphite: ["#4A4A52", "#F4F2E8"],
};

const SIZE = 1024;
const SS = 4; // supersampling factor

// Geometry measured on the original honey icon (1024x1024) — see generate-icons.py.
const CLOCK_CENTER = [415, 500];
const CLOCK_RADIUS = 268;
const RING_STROKE = 46;
const HAND_STROKE = 42;
const CENTER_DOT_RADIUS = 30;
const MINUTE_ANGLE_DEG = 48; // clockwise from 12 o'clock — 10:08 check mark
const MINUTE_LENGTH = 192;
const HOUR_ANGLE_DEG = 304;
const HOUR_LENGTH = 134;
const LINE_STROKE = 38;
const LINE_X_START = 740;
const PLANNING_LINES = [
  [432, 1005],
  [510, 925],
  [588, 850],
];

function hexToRgb(hex) {
  return [1, 3, 5].map((i) => parseInt(hex.slice(i, i + 2), 16));
}

function polarOffset([ox, oy], angleDeg, length) {
  const rad = (angleDeg * Math.PI) / 180;
  return [ox + Math.sin(rad) * length, oy - Math.cos(rad) * length];
}

function distToSegment(px, py, [ax, ay], [bx, by]) {
  const dx = bx - ax;
  const dy = by - ay;
  const lenSq = dx * dx + dy * dy;
  const t = lenSq === 0 ? 0 : Math.max(0, Math.min(1, ((px - ax) * dx + (py - ay) * dy) / lenSq));
  const cx = ax + t * dx;
  const cy = ay + t * dy;
  return Math.hypot(px - cx, py - cy);
}

/** 1 bit per supersampled pixel: is this pixel ink? */
function buildInkMask() {
  const size = SIZE * SS;
  const mask = new Uint8Array(size * size);
  const [cx, cy] = [CLOCK_CENTER[0] * SS, CLOCK_CENTER[1] * SS];
  const radius = CLOCK_RADIUS * SS;
  const ringMid = radius - (RING_STROKE * SS) / 2;
  const ringHalf = (RING_STROKE * SS) / 2;
  const handHalf = (HAND_STROKE * SS) / 2;
  const dotR = CENTER_DOT_RADIUS * SS;
  const lineHalf = (LINE_STROKE * SS) / 2;

  const minuteTip = polarOffset([cx, cy], MINUTE_ANGLE_DEG, MINUTE_LENGTH * SS);
  const hourTip = polarOffset([cx, cy], HOUR_ANGLE_DEG, HOUR_LENGTH * SS);
  const lines = PLANNING_LINES.map(([y, xEnd]) => [
    [LINE_X_START * SS, y * SS],
    [xEnd * SS, y * SS],
  ]);

  for (let y = 0; y < size; y++) {
    const py = y + 0.5;
    for (let x = 0; x < size; x++) {
      const px = x + 0.5;
      const dClock = Math.hypot(px - cx, py - cy);
      const isInk =
        Math.abs(dClock - ringMid) <= ringHalf ||
        dClock <= dotR ||
        (dClock <= radius &&
          (distToSegment(px, py, [cx, cy], minuteTip) <= handHalf ||
            distToSegment(px, py, [cx, cy], hourTip) <= handHalf)) ||
        lines.some(([a, b]) => distToSegment(px, py, a, b) <= lineHalf);
      if (isInk) mask[y * size + x] = 1;
    }
  }
  return mask;
}

/** Downsample the mask (SSxSS box average) and blend accent→ink into RGB rows. */
function rasterize(mask, accent, ink) {
  const [ar, ag, ab] = hexToRgb(accent);
  const [ir, ig, ib] = hexToRgb(ink);
  const big = SIZE * SS;
  // Each scanline is prefixed by filter byte 0.
  const raw = Buffer.alloc(SIZE * (1 + SIZE * 3));
  for (let y = 0; y < SIZE; y++) {
    const rowStart = y * (1 + SIZE * 3) + 1;
    for (let x = 0; x < SIZE; x++) {
      let cover = 0;
      for (let sy = 0; sy < SS; sy++) {
        const base = (y * SS + sy) * big + x * SS;
        for (let sx = 0; sx < SS; sx++) cover += mask[base + sx];
      }
      const t = cover / (SS * SS);
      const offset = rowStart + x * 3;
      raw[offset] = Math.round(ar + (ir - ar) * t);
      raw[offset + 1] = Math.round(ag + (ig - ag) * t);
      raw[offset + 2] = Math.round(ab + (ib - ab) * t);
    }
  }
  return raw;
}

const CRC_TABLE = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    table[n] = c >>> 0;
  }
  return table;
})();

function crc32(buffer) {
  let crc = 0xffffffff;
  for (const byte of buffer) crc = CRC_TABLE[(crc ^ byte) & 0xff] ^ (crc >>> 8);
  return (crc ^ 0xffffffff) >>> 0;
}

function pngChunk(type, data) {
  const length = Buffer.alloc(4);
  length.writeUInt32BE(data.length);
  const body = Buffer.concat([Buffer.from(type, "ascii"), data]);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(body));
  return Buffer.concat([length, body, crc]);
}

function encodePng(raw) {
  const signature = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(SIZE, 0);
  ihdr.writeUInt32BE(SIZE, 4);
  ihdr[8] = 8; // bit depth
  ihdr[9] = 2; // color type RGB
  return Buffer.concat([
    signature,
    pngChunk("IHDR", ihdr),
    pngChunk("IDAT", deflateSync(raw, { level: 9 })),
    pngChunk("IEND", Buffer.alloc(0)),
  ]);
}

mkdirSync(ICONS_DIR, { recursive: true });
const mask = buildInkMask();
for (const [themeId, [accent, ink]] of Object.entries(THEMES)) {
  const png = encodePng(rasterize(mask, accent, ink));
  const target = join(ICONS_DIR, `icon-${themeId}.png`);
  writeFileSync(target, png);
  console.log(`wrote ${target}`);
}
const [accent, ink] = THEMES.honey;
writeFileSync(DEFAULT_ICON, encodePng(rasterize(mask, accent, ink)));
console.log(`wrote ${DEFAULT_ICON} (default, honey)`);
